pkg/asset/machines/gcp/gcpmachines.go

```diff
@@ -165,8 +174,29 @@ func createGCPMachine(name string, installConfig *installconfig.InstallConfig, i
 	// value in install-config.
 	// Note - the derivation of the ServiceAccount from credentials will no longer be supported.
 	if len(installConfig.Config.Platform.GCP.NetworkProjectID) > 0 {
-		if mpool.ServiceAccount != "" {
-			serviceAccount.Email = mpool.ServiceAccount
+		serviceAccount.Email = mpool.ServiceAccount
+		if serviceAccount.Email == "" {
+			sess, err := gcpconfig.GetSession(context.TODO())
+			if err != nil {
+				return nil, fmt.Errorf("gcp machine creation failed to get session: %w", err)
+			}
+
+			// The JSON can be `nil` if auth is provided from env
+			// https://pkg.go.dev/golang.org/x/oauth2@v0.17.0/google#Credentials
+			if len(sess.Credentials.JSON) == 0 {
+				return nil, fmt.Errorf("could not extract service account from loaded credentials. Please specify a service account to be used for shared vpc installations in the install-config.yaml")
+			}
+
+			var found bool
+			sa := make(map[string]interface{})
+			err = json.Unmarshal(sess.Credentials.JSON, &sa)
+			if err != nil {
+				return nil, fmt.Errorf("failed to unmarshal gcp session: %w", err)
+			}
+			serviceAccount.Email, found = sa["client_email"].(string)
+			if !found {
+				return nil, fmt.Errorf("could not find google service account")
+			}
```

Comment:
In the future, you might want to move this outside `createGCPMachine` so it's executed only once for control plane and bootstrap machines.